File: swix/swix/swix/objc/math.m
//
//  fft.m
//  swix
//
//  Created by Scott Sievert on 7/9/14.
//  Copyright (c) 2014 com.scott. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <Accelerate/Accelerate.h>

void svd_objc(double * xx, int m, int n, double* s, double* vt, double* u, int compute_uv){
    // adapted from the buggy code at http://stackoverflow.com/questions/5047503/lapack-svd-singular-value-decomposition
    
    char job;
    if      (compute_uv == 0) job = 'N';
    else if (compute_uv == 1) job = 'A';
    
    __CLPK_integer lda = (__CLPK_integer)m;
    long numberOfSingularValues = m < n ? m : n;
    
    // Workspace and status variables:
    __CLPK_integer _n = n;
    __CLPK_integer _m = m;
    double* work = (double*)malloc(sizeof(double) * 2);
    __CLPK_integer lwork = -1;
    __CLPK_integer * iwork = (__CLPK_integer *)malloc(sizeof(__CLPK_integer) * 8 * numberOfSingularValues);
    __CLPK_integer info = 0;
    
    // Call dgesdd_ with lwork = -1 to query optimal workspace size:
    dgesdd_(&job, &_m, &_n, xx, &lda, s, u, &_m, vt, &_n, work, &lwork, iwork, &info);
    
    // Optimal workspace size is returned in work[0].
    lwork = work[0];
    free(work);
    work = (double *)malloc(lwork * sizeof(double));
    
    // Call dgesdd_ to do the actual computation:
    dgesdd_(&job, &_m, &_n, xx, &lda, s, u, &_m, vt, &_n, work, &lwork, iwork, &info);
    
    free(work);
    free(iwork);
}

File: swix/swix/swix/math.swift
//
//  math.swift
//  swix
//
//  Created by Scott Sievert on 7/11/14.
//  Copyright (c) 2014 com.scott. All rights reserved.
//

import Foundation
import Accelerate

// fft, ifft, dot product, haar wavelet
func dot(x: matrix2d, y: matrix2d) -> matrix2d{
    var (Mx, Nx) = x.shape
    var (My, Ny) = y.shape
    assert(Nx == My, "Matrix sizes not compatible for dot product")
    var z = zeros((Mx, Ny))
    
    dot_objc(!x, !y, !z, Mx.cint, Ny.cint, Nx.cint)
    
    return z
}


func fft(x: ndarray) -> (ndarray, ndarray){
    var N:CInt = x.n.cint
    var yr = zeros(N.int)
    var yi = zeros(N.int)
    fft_objc(!x, N, !yr, !yi);
    
    return (yr, yi)
}
func ifft(yr: ndarray, yi: ndarray) -> ndarray{
    var N = yr.n
    var x = zeros(N)
    ifft_objc(!yr, !yi, N.cint, !x);

    return x
}